opm/simulators/wells/MultisegmentWellSegments.cpp

```diff
@@ -583,17 +584,39 @@ pressureDropSpiralICD(const int seg) const
     }
 
     EvalWell density = densities_[seg_upwind];
-    // WARNING
-    // We disregard the derivatives from the upwind density to make sure derivatives
-    // wrt. to different segments dont get mixed.
+    EvalWell mass_rate = mass_rates_[seg];
+    // In the reverse flow case, we don't have enough slots for all derivatives, e.g.,
+    // upwind pressure and flow. We amend this by a second function call option, where 
+    // only these remaining derivatives are considered.
+    // For reference: the pressure equation assumes pressure/flow derivatives are given 
+    // at segment node while fraction derivatives are given at upwind node.   
     if (seg != seg_upwind) {
-        water_fraction.clearDerivatives();
-        water_viscosity.clearDerivatives();
-        oil_fraction.clearDerivatives();
-        oil_viscosity.clearDerivatives();
-        gas_fraction.clearDerivatives();
-        gas_viscosity.clearDerivatives();
-        density.clearDerivatives();
+        constexpr int nvar = FluidSystem::numPhases + 1;
+        std::vector<bool> zero_mask(nvar, false);
+        if (!return_extra_derivatives){
+            zero_mask[PrimaryVariables::WQTotal] = true;
+            zero_mask[PrimaryVariables::SPres] = true;
+        } else {
+            if (PrimaryVariables::has_water){
+                zero_mask[PrimaryVariables::WFrac] = true;
+            }
+            if (PrimaryVariables::has_gas){
+                zero_mask[PrimaryVariables::GFrac] = true;
+            }
```

Comment:
These could probably be
```C++
if constexpr (PrimaryVariables::has_X) {
    // ...
}
```
instead, but I don't know that marking them as such would give any particular benefit here.  The compiler should be able to infer that the conditions [are constant](https://github.com/OPM/opm-simulators/blob/b60eb25285cc2cf1a78376bb64249e6af7d0ed7c/opm/simulators/wells/MultisegmentWellPrimaryVariables.hpp#L59-L61) anyway.